### PR TITLE
Added TCP healthcheck capabilities to the HdsDelegate

### PR DESCRIPTION
**Pull request by markatou**
The HdsDelegate now informs the management server that it can perform TCP healthchecks. 

This change also comes with three tests showing that indeed the HdsDelegate can perform TCP healthchecks. 

*Risk Level*: Low

This is for https://github.com/envoyproxy/envoy/issues/1310.

### CODE REVIEW

**Comment by htuch** [comment]
@hennna can you take a pass on this? Thanks.

**Comment by markatou** [comment]
@hennna friendly ping

File: test/integration/hds_integration_test.cc [hunk 1]
```diff
@@ -145,6 +145,40 @@ class HdsIntegrationTest : public HttpIntegrationTest,
     return server_health_check_specifier_;
   }
 
+  // Creates a basic HealthCheckSpecifier message containing one endpoint and
+  // one TCP health_check
+  envoy::service::discovery::v2::HealthCheckSpecifier makeTCPHealthCheckSpecifier() {
```
Comment (htuch) [inline test/integration/hds_integration_test.cc]: Nit: `makeTcpHealthCheckSpecifier`

File: test/integration/hds_integration_test.cc [hunk 2]
```diff
@@ -145,6 +145,40 @@ class HdsIntegrationTest : public HttpIntegrationTest,
...
+        ->set_port_value(host_upstream_->localAddress()->ip()->port());
+    health_check->mutable_endpoints(0)->mutable_locality()->set_region("some_region");
+    health_check->mutable_endpoints(0)->mutable_locality()->set_zone("some_zone");
+    health_check->mutable_endpoints(0)->mutable_locality()->set_sub_zone("crete");
+
+    health_check->add_health_checks()->mutable_timeout()->set_seconds(1);
+    health_check->mutable_health_checks(0)->mutable_interval()->set_seconds(1);
+    health_check->mutable_health_checks(0)->mutable_unhealthy_threshold()->set_value(2);
+    health_check->mutable_health_checks(0)->mutable_healthy_threshold()->set_value(2);
+    auto tcp_hc = health_check->mutable_health_checks(0)->mutable_tcp_health_check();
```
Comment (htuch) [inline test/integration/hds_integration_test.cc]: Nit: `auto*`

File: test/integration/hds_integration_test.cc [hunk 3]
```diff
@@ -145,6 +145,40 @@ class HdsIntegrationTest : public HttpIntegrationTest,
...
+    health_check->add_endpoints()
+        ->add_endpoints()
+        ->mutable_address()
+        ->mutable_socket_address()
+        ->set_address(host_upstream_->localAddress()->ip()->addressAsString());
+    health_check->mutable_endpoints(0)
+        ->mutable_endpoints(0)
+        ->mutable_address()
+        ->mutable_socket_address()
+        ->set_port_value(host_upstream_->localAddress()->ip()->port());
```
Comment (htuch) [inline test/integration/hds_integration_test.cc]: Nit: consider using `addressToProtobufAddress()` in `source/common/network/utility.h` for this (and elsewhere in the HDS tests).

File: test/integration/hds_integration_test.cc [hunk 4]
```diff
@@ -293,12 +330,120 @@ TEST_P(HdsIntegrationTest, SingleEndpointUnhealthy) {
   cleanupHdsConnection();
 }
 
-// Tests that Envoy can healthcheck two hosts that are in the same cluster, and
+// Tests Envoy TCP healthchecking an endpoint that doesn't respond and reporting that it is
```
Comment (htuch) [inline test/integration/hds_integration_test.cc]: Nit: `health checking`

File: test/integration/hds_integration_test.cc [hunk 5]
```diff
@@ -293,12 +330,120 @@ TEST_P(HdsIntegrationTest, SingleEndpointUnhealthy) {
...
+  server_health_check_specifier_ = makeTCPHealthCheckSpecifier();
+  hds_stream_->startGrpcStream();
+  hds_stream_->sendGrpcMessage(server_health_check_specifier_);
+  test_server_->waitForCounterGe("hds_delegate.requests", ++hds_requests_);
+
+  // Envoy asks the endpoint
+  ASSERT_TRUE(host_upstream_->waitForRawConnection(host_fake_raw_connection_));
+  ASSERT_TRUE(
+      host_fake_raw_connection_->waitForData(FakeRawConnection::waitForInexactMatch("Ping")));
+  auto ret = host_fake_raw_connection_->write("Voronoi");
```
Comment (htuch) [inline test/integration/hds_integration_test.cc]: I like the references..

File: test/integration/hds_integration_test.cc [hunk 6]
```diff
@@ -293,12 +330,120 @@ TEST_P(HdsIntegrationTest, SingleEndpointUnhealthy) {
...
+  server_health_check_specifier_ = makeTCPHealthCheckSpecifier();
+  hds_stream_->startGrpcStream();
+  hds_stream_->sendGrpcMessage(server_health_check_specifier_);
+  test_server_->waitForCounterGe("hds_delegate.requests", ++hds_requests_);
+
+  // Envoy asks the endpoint
+  ASSERT_TRUE(host_upstream_->waitForRawConnection(host_fake_raw_connection_));
+  ASSERT_TRUE(
+      host_fake_raw_connection_->waitForData(FakeRawConnection::waitForInexactMatch("Ping")));
+  auto ret = host_fake_raw_connection_->write("Pong");
```
Comment (htuch) [inline test/integration/hds_integration_test.cc]: Nit: is `ref` unused below?

File: test/integration/hds_integration_test.cc [hunk 7]
```diff
@@ -295,30 +316,140 @@ TEST_P(HdsIntegrationTest, SingleEndpointUnhealthy) {
...
+  EXPECT_EQ(envoy_msg_.capability().health_check_protocol(1),
+            envoy::service::discovery::v2::Capability::TCP);
+
+  // Server asks for health checking
+  server_health_check_specifier_ = makeTcpHealthCheckSpecifier();
+  hds_stream_->startGrpcStream();
+  hds_stream_->sendGrpcMessage(server_health_check_specifier_);
+  test_server_->waitForCounterGe("hds_delegate.requests", ++hds_requests_);
+
+  // Envoys asks the endpoint
```
Comment (hennna) [inline test/integration/hds_integration_test.cc]: comment nit: what is Envoy asking for?

**Comment by htuch** [comment]
@markatou also need to resolve merge conflict :)

**Pull request merged**